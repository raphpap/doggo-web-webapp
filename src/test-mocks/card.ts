// Types
import {Card} from 'doggo/context';

export const mockCard = (id: number): Card => ({
  attack: id,
  defense: id,
  hpLeft: id,
  hpTotal: id,
  id: `${id}`,
  image: `${id}`,
  name: `${id}`,
});
