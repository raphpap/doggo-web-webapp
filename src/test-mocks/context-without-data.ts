// Types
import {ApplicationContext, BattleStatus} from 'doggo/context';

export const mockContextWithoutData = (): ApplicationContext => ({
  actions: {
    battle: jest.fn(),
    capture: jest.fn(),
    getNextOpponent: jest.fn(),
    login: jest.fn(),
    selectBattleCard: jest.fn(),
    unselectBattleCard: jest.fn(),
  },
  state: {
    battle: {
      cardId: null,
      opponent: null,
      status: BattleStatus.NoOpponent
    },
    cards: null,
    error: null,
    loading: false
  }
});
