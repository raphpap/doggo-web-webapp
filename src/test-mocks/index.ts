export * from './card';
export * from './context-without-data';
export * from './logged-in';
