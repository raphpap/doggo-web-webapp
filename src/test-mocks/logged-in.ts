// Types
import {ApplicationContext, BattleStatus} from 'doggo/context';

// Mocks
import {mockCard, mockContextWithoutData} from '.';

export const mockLoggedIn = (): ApplicationContext => {
  const context = mockContextWithoutData();

  return {
    ...context,
    state: {
      ...context.state,
      battle: {
        ...context.state.battle,
        opponent: mockCard(99),
        status: BattleStatus.NoCardSelected
      },
      cards: [
        mockCard(1),
        mockCard(2)
      ]
    }
  };
};
