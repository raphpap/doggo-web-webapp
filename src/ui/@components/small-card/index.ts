export * from './small-card';
export {default} from './small-card';
