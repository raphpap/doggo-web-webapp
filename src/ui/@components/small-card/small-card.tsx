// Vendor
import React from 'react';
import styled from 'react-emotion';

// Context
import {Card as CardType} from 'doggo-web-webapp/ui/@context';

// Elements
const CardContainer = styled.li`
  display: flex;
  justify-content: space-between;
  width: 400px;
  height: 80px;
  border: 1px solid rgba(255, 255, 255, 0.8);
  margin: 8px;
  cursor: pointer;
`;

const Image = styled.img`
  display: block;
  width: auto;
  height: 100%;
`;

const Info = styled.div`
  display: flex;
  flex: 1 1 auto;
  flex-direction: column;
  justify-content: space-around;
  padding: 8px 24px;
  color: rgba(255, 255, 255, 0.8);
`;

const Stats = styled.div`
  color: rgba(255, 255, 255, 0.6);
`;

// Types
interface Props {
  card: CardType;
  onCardClick: (card: CardType) => void;
}

export const SmallCard: React.SFC<Props> = ({card, onCardClick}) => {
  const {attack, defense, hp, name, image} = card;

  return (
    <CardContainer onClick={() => {onCardClick(card)}}>
      <Image src={image} />

      <Info>
        <div>{name}</div>
        <Stats>{hp}hp {attack}atk {defense}def</Stats>
      </Info>
    </CardContainer>
  );
};

export default SmallCard;
