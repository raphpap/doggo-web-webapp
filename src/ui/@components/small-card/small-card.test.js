// Vendor
import React from 'react';
import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';

// Components
import {CardContainer, SmallCard} from './small-card';

// Mocks
import {mockCard} from 'doggo/test-mocks';

describe('<SmallCard />', () => {
  let wrapper;
  let props;
  let card;

  beforeEach(() => {
    card = {
      attack: 10,
      defense: 20,
      hpLeft: 30,
      hpTotal: 40,
      id: '123',
      image: 'image.jpg',
      name: 'card_name'
    };

    props = {
      card,
      isNotClickable: false,
      onCardClick: jest.fn()
    };

    wrapper = shallow(<SmallCard {...props} />);
  });

  it('renders correctly', () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it('renders CardContainer with the right props when clickable', () => {
    const cardContainer = wrapper.find(CardContainer);
    expect(cardContainer.props().isClickable).toBe(true);
  });

  it('renders CardContainer with the right props when not clickable', () => {
    props = {
      ...props,
      isNotClickable: true,
    };
    wrapper.setProps(props);

    const cardContainer = wrapper.find(CardContainer);
    expect(cardContainer.props().isClickable).toBe(false);
  });


  it('renders CardContainer with the right props when some hp is left', () => {
    const cardContainer = wrapper.find(CardContainer);
    expect(cardContainer.props().isDefeated).toBe(false);
  });

  it('renders CardContainer with the right props when no hp is left', () => {
    props = {
      ...props,
      card: {
        ...props.card,
        hpLeft: 0
      },
    };
    wrapper.setProps(props);

    const cardContainer = wrapper.find(CardContainer);
    expect(cardContainer.props().isDefeated).toBe(true);
  });

  it('calls onCardClick when CardContainer is clicked while being clickable', () => {
    const cardContainer = wrapper.find(CardContainer);
    cardContainer.simulate('click');

    expect(props.onCardClick).toHaveBeenCalledWith(props.card);
  });

  it('does not calls onCardClick when CardContainer is clicked while not being clickable', () => {
    props = {
      ...props,
      isNotClickable: true
    };
    wrapper.setProps(props);

    const cardContainer = wrapper.find(CardContainer);
    cardContainer.simulate('click');

    expect(props.onCardClick).toHaveBeenCalledTimes(0);
  });
});
