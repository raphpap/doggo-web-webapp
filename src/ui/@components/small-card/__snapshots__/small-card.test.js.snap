// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SmallCard /> renders correctly 1`] = `
<CardContainer
  isClickable={true}
  isDefeated={false}
  onClick={[Function]}
>
  <Image
    src="image.jpg"
  />
  <Info>
    <Name>
      card_name
    </Name>
    <HealthBar
      hpLeft={30}
      hpTotal={40}
    />
    <Stats>
      <div>
        Atk: 
        10
      </div>
      <div>
        Def: 
        20
      </div>
    </Stats>
  </Info>
</CardContainer>
`;
